kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala

```diff
@@ -205,6 +205,18 @@ private[kyuubi] class EngineRef(
             throw error
           }
         }
+
+        engineManager.getApplicationInfo(builder.clusterManager(), engineRefId).foreach { appInfo =>
+          if (ApplicationState.isTerminated(appInfo.state)) {
+            throw new KyuubiSQLException(
+              s"""
+                 |The engine application has been in terminate state. Please check the engine log.
+                 |FYI: ${appInfo.toMap.mkString("(\n", ",\n", "\n)")}
+                 |""".stripMargin,
+              builder.getError)
+          }
+        }
+
```

Comment:
I can not get it, what's wrong if we kill a terminated application ? And here should close the submit `process` ?

Reply:
If the engine is terminate，we shall not wait the engine available.

Reply:
Yes， we need close the submit process.

Reply:
the problem is: the submit process is available but the application is terminated ? can you describe how it happened

Reply:
FYI:
The engineId is `83bdadc9-88df-4c6a-b700-dd1e07dffb65`.

<img width="1500" alt="image" src="https://user-images.githubusercontent.com/6757692/186299228-c5689f2d-6836-4f62-8384-e4447b4fe108.png">

and the submit process finished at 1:46:19

<img width="1380" alt="image" src="https://user-images.githubusercontent.com/6757692/186299278-f1cf5d7d-0722-4a35-a5b2-2f146985d7d5.png">

But kyuubi close the session until the init timeout.
<img width="1493" alt="image" src="https://user-images.githubusercontent.com/6757692/186299386-218046b2-ebf1-41f8-a3a4-09ec8dacff1d.png">

Reply:
I see this issue, it might be some problem between spark engine initialization and register itself to discovery service when the submit process does not affect the application status (i.e. cluster mode).

In this case, shall we put the application info status check code inside the `exitValue == 0` branch ? this issue should only happen if process exit successfully.

Reply:
we can not rely on the exitValue of spark-submit process.

Maybe we do not wait the app completion and the spark-submit process terminate once the app submitted to resource manager.

Reply:
> Maybe we do not wait the app completion and the spark-submit process terminate once the app submitted to resource manager.

in this case, the exitValue of submit process is 0 ..

Reply:
@turboFei this comment has not resovled.

There are three kinds of status of submit process:
- not exit
- exit normally with code == 0
- exit with code != 0

Do we nned check the status from application manager with option 1 and 3 ?

For option 1, if the submit process is running how can the status from application manager become terminated ? if so, it should be a bug from engine (e.g. spark).
For option 3, we have already fast failed so it does not affect.

Reply:
Seems for 1, it is not needed.

So you mean that, only check the app terminate on case 2.

Reply:
@ulysses-you thanks, resolved.